File: backend/app/executors/web_source_executor.py
"""
WebSource Node Executor

Executes web scraping operations using headless browser technology.
Handles visual selector configurations and structured data extraction.
"""

import asyncio
from typing import Dict, Any, Optional, List
from .base_executor import BaseExecutor
from app.services.playwright_service import PlaywrightService

class WebSourceExecutor(BaseExecutor):
    """Executor for WebSource nodes using headless browser scraping"""
    
    async def execute(self, input_data: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """
        Execute web scraping using Playwright-based headless browser
        
        Expected node configuration:
        {
            "id": "node_id",
            "type": "webSource",
            "config": {
                "url": "https://example.com",
                "selectors": [
                    {
                        "name": "title",
                        "selector": "h1",
                        "type": "all|table|repeating",
                        "attribute": "textContent",
                        "fields": [...] // for repeating type
                    }
                ],
                "options": {
                    "wait_for_load": true,
                    "timeout": 30000
                }
            }
        }
        """
        
        self.log_execution_start()
        
        try:
            # Validate configuration
            if not self.validate_config():
                return self.create_error_result("Invalid node configuration")
            
            # Extract configuration
            config = self.get_config_value('config', {})
            url = config.get('url')
            selectors = config.get('selectors', [])
            options = config.get('options', {})
            
            # Validate required fields
            if not url:
                return self.create_error_result("Missing required field: url")
            
            if not selectors:
                return self.create_error_result("Missing required field: selectors")
            
            # Set default options
            wait_for_load = options.get('wait_for_load', True)
            timeout = options.get('timeout', 30000)
            
            self.logger.info(f"Scraping URL: {url}")
            self.logger.info(f"Processing {len(selectors)} selectors")
            
            # Perform the scraping
            result = await self._scrape_with_playwright(url, selectors, wait_for_load, timeout)
            
            if result['success']:
                self.log_execution_complete(f"Extracted data from {len(selectors)} selectors")
                return self.create_success_result(
                    data=result['data'],
                    metadata={
                        'url': url,
                        'page_info': result.get('page_info', {}),
                        'selectors_processed': result.get('selectors_processed', 0),
                        'extraction_method': 'headless_browser'
                    }
                )
            else:
                self.log_execution_error(Exception(result.get('error', 'Unknown error')))
                return self.create_error_result(
                    error=result.get('error', 'Scraping failed'),
                    details=result
                )
                
        except Exception as e:
            self.log_execution_error(e)
            return self.create_error_result(f"Execution failed: {str(e)}")
    
    async def _scrape_with_playwright(self, url: str, selectors: List[Dict], wait_for_load: bool, timeout: int) -> Dict[str, Any]:
        """
        Perform the actual scraping using PlaywrightService
        """
        async with PlaywrightService() as service:
            try:
                # Load the page
                self.logger.info(f"Loading page: {url}")
                await service.load_page(url, wait_for_load)
                
                # Get page info
                page_info = await service.get_page_info()
                self.logger.info(f"Page loaded: {page_info.get('title', 'Unknown')}")
                
                # Process selectors to ensure they're in the correct format
                processed_selectors = self._process_selectors(selectors)
                
                # Extract data using the enhanced extraction method
                extracted_data = await service.extract_all_content(processed_selectors)
                
                self.logger.info(f"Extraction completed. Results: {len(extracted_data) if extracted_data else 0} items")
                
                return {
                    'success': True,
                    'url': url,
                    'data': extracted_data[0] if extracted_data else {},
                    'page_info': page_info,
                    'selectors_processed': len(selectors)
                }
                
            except Exception as e:
                self.logger.error(f"Error scraping {url}: {e}")
                return {
                    'success': False,
                    'url': url,
                    'error': str(e),
                    'data': {},
                    'page_info': {},
                    'selectors_processed': 0
                }
    
    def _process_selectors(self, selectors: List[Dict]) -> List[Dict]:
        """
        Process and validate selector configurations
        """
        processed = []
        
        for selector_config in selectors:
            if not isinstance(selector_config, dict):
                self.logger.warning(f"Invalid selector config (not dict): {selector_config}")
                continue
            
            # Ensure required fields
            if 'selector' not in selector_config:
                self.logger.warning(f"Selector config missing 'selector' field: {selector_config}")
                continue
            
            # Set defaults
            processed_config = {
                'selector': selector_config['selector'],
                'name': selector_config.get('name', selector_config.get('label', selector_config['selector'])),
                'type': selector_config.get('type', 'all'),
                'attribute': selector_config.get('attribute', 'textContent')
            }
            
            # Handle repeating containers with fields
            if processed_config['type'] == 'repeating' and 'fields' in selector_config:
                processed_config['fields'] = selector_config['fields']
            
            # Handle table extraction
            if processed_config['type'] == 'table':
                processed_config['attribute'] = 'table_data'
            
            processed.append(processed_config)
            self.logger.debug(f"Processed selector: {processed_config}")
        
        return processed
    
    def validate_config(self) -> bool:
        """
        Validate WebSource-specific configuration
        """
        if not super().validate_config():
            return False
        
        config = self.get_config_value('config', {})
        
        # Check for required configuration fields
        if not config.get('url'):
            self.logger.error("WebSource node missing 'url' in config")
            return False
        
        selectors = config.get('selectors', [])
        if not selectors or not isinstance(selectors, list):
            self.logger.error("WebSource node missing valid 'selectors' array in config")
            return False
        
        # Validate each selector
        for i, selector in enumerate(selectors):
            if not isinstance(selector, dict):
                self.logger.error(f"Selector {i} is not a dictionary")
                return False
            
            if not selector.get('selector'):
                self.logger.error(f"Selector {i} missing 'selector' field")
                return False
        
        return True

File: services/workflow-frontend/src/components/workflow-builder/execution/WebSourceExecutor.ts
import { ExecutionContext, NodeOutput, WorkflowNodeData, DataType } from './types';
import { ParameterGenerator } from './ParameterGenerator';
import { workflowService } from '../services';

export class WebSourceExecutor {
  private parameterGenerator: ParameterGenerator;

  constructor(private context: ExecutionContext) {
    this.parameterGenerator = new ParameterGenerator();
  }

  /**
   * Execute web source node
   */
  async execute(nodeId: string, nodeData: WorkflowNodeData, inputs: NodeOutput[]): Promise<any> {
    const { monitoring } = nodeData;
    console.log('🔍 DEBUG: nodeData:', nodeData);
    console.log('🔍 DEBUG: monitoring config:', monitoring);
    
    if (!monitoring) {
      throw new Error('Web source not configured');
    }

    // Check for loop configuration
    if (monitoring.loopConfig?.enabled && monitoring.loopConfig.parameters.length > 0) {
      return await this.executeWithLoop(nodeId, monitoring, inputs);
    }

    return await this.executeSingle(nodeId, monitoring, inputs);
  }

  /**
   * Execute single web source extraction
   */
  private async executeSingle(nodeId: string, monitoring: any, inputs: NodeOutput[]): Promise<any> {
    // Create workflow execution if it doesn't exist
    const executionResponse = await fetch(`${import.meta.env.VITE_API_BASE_URL}/api/v1/agents/${this.context.agentId}/executions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.context.authHeaders
      },
      body: JSON.stringify({
        workflow_nodes: this.context.nodes,
        workflow_edges: this.context.edges
      })
    });

    if (!executionResponse.ok) {
      throw new Error('Failed to create workflow execution');
    }

    const executionData = await executionResponse.json();
    const executionId = executionData.execution?.id;

    // Create node execution record for tracking
    const nodeExecutionResponse = await fetch(`${import.meta.env.VITE_API_BASE_URL}/api/v1/executions/${executionId}/nodes`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.context.authHeaders
      },
      body: JSON.stringify({
        node_id: nodeId,
        node_type: 'webSource',
        input_config: {
          nodeType: 'webSource',
          url: monitoring.url,
          selectors: monitoring.selectors
        },
        input_data: inputs.map(input => input.data),
        status: 'running'
      })
    });

    if (!nodeExecutionResponse.ok) {
      throw new Error('Failed to create node execution record');
    }

    const nodeExecutionData = await nodeExecutionResponse.json();
    const nodeExecutionId = nodeExecutionData.node_execution?.id;
    
    // Use the imported workflowService
    
    // Always prefer originalSelectors if available - they have the correct structure
    const selectors = monitoring.originalSelectors || monitoring.selectors || [];
    
    console.log('🔍 SELECTOR SOURCE:', monitoring.originalSelectors ? 'Using originalSelectors' : 'Using selectors');
    console.log('🔍 MONITORING OBJECT:', JSON.stringify(monitoring, null, 2));
    console.log('🔍 SELECTORS:', JSON.stringify(selectors, null, 2));
    
    console.log('🔍 DETAILED SELECTORS DEBUG:', JSON.stringify(selectors, null, 2));
    
    // Enhanced format detection logging
    if (selectors.length > 0) {
      const firstSelector = selectors[0];
      console.log('🔍 FORMAT DETECTION ANALYSIS:');
      console.log('  - First selector has .selector:', !!firstSelector.selector);
      console.log('  - First selector has .id:', !!firstSelector.id);
      console.log('  - Detection condition (selector && !id):', !!(firstSelector.selector && !firstSelector.id));
      console.log('  - First selector type:', firstSelector.type);
      
      // Check for repeating selectors specifically
      const repeatingSelector = selectors.find((sel: any) => sel.type === 'repeating');
      if (repeatingSelector) {
        console.log('🔍 REPEATING SELECTOR ANALYSIS:');
        console.log('  - Found repeating selector:', !!repeatingSelector);
        console.log('  - Has fields property:', !!repeatingSelector.fields);
        console.log('  - Fields count:', repeatingSelector.fields?.length || 0);
        if (repeatingSelector.fields?.length > 0) {
          console.log('  - Sample field structure:', JSON.stringify(repeatingSelector.fields[0], null, 2));
          repeatingSelector.fields.forEach((field: any, idx: number) => {
            console.log(`  - Field ${idx}: name="${field.name}", sub_selector="${field.sub_selector}"`);
          });
        }
      }
    }
    
    // Send raw selector configuration directly to backend - let backend handle extraction logic
    console.log('🚀 WORKFLOW EXECUTION - SENDING RAW CONFIG TO BACKEND:', { 
      url: monitoring.url, 
      selectors: selectors
    });
    
    try {
      const result = await workflowService.extractData(
        monitoring.url,
        selectors,
        this.context.authHeaders
      );
      
      console.log('🎯 Workflow extraction result:', result);

      if (!result.success) {
        throw new Error(result.error || 'Failed to extract data');
      }
      
      const extractedData = result.result?.extracted_data || [];
      console.log('📥 WORKFLOW EXECUTION - EXTRACTED DATA:', extractedData);

      // Note: Node execution status updates temporarily disabled due to API method not supported
      // TODO: Implement proper execution tracking endpoint
      // await fetch(`${import.meta.env.VITE_API_BASE_URL}/api/v1/executions/${executionId}/nodes/${nodeExecutionId}`, {
      //   method: 'PATCH',
      //   headers: {
      //     'Content-Type': 'application/json',
      //     ...this.context.authHeaders
      //   },
      //   body: JSON.stringify({
      //     status: 'completed',
      //     output_data: extractedData,
      //     completed_at: new Date().toISOString()
      //   })
      // });

      return extractedData;
    } catch (error: any) {
      console.error('🚨 Workflow extraction failed:', error);
      
      // Note: Node execution status updates temporarily disabled due to API method not supported
      // TODO: Implement proper execution tracking endpoint for errors
      // await fetch(`${import.meta.env.VITE_API_BASE_URL}/api/v1/executions/${executionId}/nodes/${nodeExecutionId}`, {
      //   method: 'PATCH',
      //   headers: {
      //     'Content-Type': 'application/json',
      //     ...this.context.authHeaders
      //   },
      //   body: JSON.stringify({
      //     status: 'failed',
      //     error_message: error.message || 'Unknown error',
      //     completed_at: new Date().toISOString()
      //   })
      // });

      throw error;
    }
  }

  /**
   * Execute web source with parameter looping
   */
  private async executeWithLoop(nodeId: string, monitoring: any, inputs: NodeOutput[]): Promise<any> {
    const startTime = Date.now();
    console.log('🔄 Starting webSource loop execution');

    // Generate parameter combinations
    const parameterCombinations = await this.parameterGenerator.generateParameterCombinations(monitoring.loopConfig, inputs);
    console.log(`🔄 Generated ${parameterCombinations.length} parameter combinations`);

    // Create workflow execution if it doesn't exist
    const executionResponse = await fetch(`${import.meta.env.VITE_API_BASE_URL}/api/v1/agents/${this.context.agentId}/executions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.context.authHeaders
      },
      body: JSON.stringify({
        workflow_nodes: this.context.nodes,
        workflow_edges: this.context.edges
      })
    });

    if (!executionResponse.ok) {
      throw new Error('Failed to create workflow execution for loop');
    }

    const executionData = await executionResponse.json();
    const executionId = executionData.execution?.id;

    // Create node execution record for tracking
    const nodeExecutionResponse = await fetch(`${import.meta.env.VITE_API_BASE_URL}/api/v1/executions/${executionId}/nodes`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.context.authHeaders
      },
      body: JSON.stringify({
        node_id: nodeId,
        node_type: 'webSource',
        input_config: {
          nodeType: 'webSource',
          url: monitoring.url,
          loopConfig: monitoring.loopConfig
        },
        url: monitoring.url,
        selectors: monitoring.originalSelectors || monitoring.selectors,
        parameters: parameterCombinations,
        status: 'running'
      })
    });

    if (!nodeExecutionResponse.ok) {
      throw new Error('Failed to create node execution record for loop');
    }

    const nodeExecutionData = await nodeExecutionResponse.json();
    const nodeExecutionId = nodeExecutionData.node_execution?.id;

    try {
      // Use the imported workflowService
      
      const results = [];
      const selectors = monitoring.originalSelectors || monitoring.selectors || [];

      console.log('🔄 Processing loop iterations...');
      
      // Execute requests in batches based on concurrency setting
      const batchSize = monitoring.loopConfig.concurrency || 1;
      const batches = [];
      
      for (let i = 0; i < parameterCombinations.length; i += batchSize) {
        batches.push(parameterCombinations.slice(i, i + batchSize));
      }

      for (let batchIndex = 0; batchIndex < batches.length; batchIndex++) {
        const batch = batches[batchIndex];
        console.log(`🔄 Processing batch ${batchIndex + 1}/${batches.length} (${batch.length} items)`);

        const batchPromises = batch.map(async (parameterSet, paramIndex) => {
          try {
            // Replace parameters in URL and selectors
            const processedUrl = this.parameterGenerator.replaceParametersInString(monitoring.url, parameterSet);
            const processedSelectors = this.parameterGenerator.replaceParametersInObject(selectors, parameterSet);

            console.log(`🔄 Iteration ${batchIndex * batchSize + paramIndex + 1}: URL=${processedUrl}, params=${JSON.stringify(parameterSet)}`);

            const result = await workflowService.extractData(
              processedUrl,
              processedSelectors,
              this.context.authHeaders
            );

            if (!result.success) {
              throw new Error(result.error || 'Failed to extract data');
            }
            
            const extractedData = result.result?.extracted_data || [];

            return {
              success: true,
              data: extractedData,
              parameters: parameterSet,
              url: processedUrl
            };
          } catch (error: any) {
            console.error(`🚨 Iteration ${batchIndex * batchSize + paramIndex + 1} failed:`, error);
            
            if (monitoring.loopConfig.stopOnError) {
              throw error;
            }
            
            return {
              success: false,
              error: error.message || 'Unknown error',
              parameters: parameterSet,
              url: this.parameterGenerator.replaceParametersInString(monitoring.url, parameterSet)
            };
          }
        });

        const batchResults = await Promise.all(batchPromises);
        results.push(...batchResults);

        // Add delay between batches if configured
        if (batchIndex < batches.length - 1 && monitoring.loopConfig.delayBetweenRequests > 0) {
          console.log(`⏱️ Waiting ${monitoring.loopConfig.delayBetweenRequests}ms before next batch...`);
          await new Promise(resolve => setTimeout(resolve, monitoring.loopConfig.delayBetweenRequests));
        }
      }

      // Aggregate results
      const aggregatedResults = this.aggregateLoopResults(results, monitoring.loopConfig);
      
      console.log(`✅ WebSource loop execution completed: ${results.length}/${parameterCombinations.length} iterations in ${Date.now() - startTime}ms`);

      // Update node execution status
      await fetch(`${import.meta.env.VITE_API_BASE_URL}/api/v1/executions/${executionId}/nodes/${nodeExecutionId}`, {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/json',
          ...this.context.authHeaders
        },
        body: JSON.stringify({
          status: 'completed',
          output_data: aggregatedResults.data,
          completed_at: new Date().toISOString()
        })
      });

      return aggregatedResults.data;
    } catch (error: any) {
      console.error('🚨 WebSource loop execution failed:', error);
      
      // Note: Node execution status updates temporarily disabled due to API method not supported
      // TODO: Implement proper execution tracking endpoint for errors
      // await fetch(`${import.meta.env.VITE_API_BASE_URL}/api/v1/executions/${executionId}/nodes/${nodeExecutionId}`, {
      //   method: 'PATCH',
      //   headers: {
      //     'Content-Type': 'application/json',
      //     ...this.context.authHeaders
      //   },
      //   body: JSON.stringify({
      //     status: 'failed',
      //     error_message: error.message || 'Unknown error',
      //     completed_at: new Date().toISOString()
      //   })
      // });

      throw error;
    }
  }

  /**
   * Process selectors for API call
   */
  // Removed processSelectors method - now sending raw config directly to backend

  /**
   * Aggregate loop execution results
   */
  private aggregateLoopResults(results: any[], loopConfig: any): { data: any } {
    const successfulResults = results.filter(r => r.success);
    
    if (successfulResults.length === 0) {
      return { data: [] };
    }

    // For webSource, we need to return just the data array, not the wrapper object
    let finalData;
    
    if (loopConfig.aggregationMode === 'merge') {
      // For merge mode with webSource data, we need special handling for named tables
      const allDataArrays = successfulResults.map(r => r.data);
      console.log(`🔄 Merge mode: Processing ${allDataArrays.length} data arrays`);
      
      // Check if all results have the same structure (array with single object containing named tables)
      if (allDataArrays.every(arr => Array.isArray(arr) && arr.length === 1 && typeof arr[0] === 'object')) {
        const mergedTables: any = {};
        
        for (const dataArray of allDataArrays) {
          const dataObject = dataArray[0];
          for (const [tableName, tableData] of Object.entries(dataObject)) {
            if (!mergedTables[tableName]) {
              mergedTables[tableName] = [];
            }
            if (Array.isArray(tableData)) {
              mergedTables[tableName].push(...tableData);
            }
          }
        }
        
        finalData = [mergedTables];
        console.log(`🔗 Merge mode: Combined data into ${Object.keys(mergedTables).length} table(s)`);
      } else {
        // Fallback: just concatenate all data
        finalData = allDataArrays.flat();
        console.log(`🔗 Merge mode fallback: Flattened ${allDataArrays.length} arrays`);
      }
    } else {
      // Append mode: extract and flatten the data arrays
      finalData = successfulResults.map(r => r.data).flat();
      console.log(`📎 Append mode: Keeping ${finalData.length} separate results`);
    }
    
    console.log(`🎯 Final webSource data:`, {
      mode: loopConfig.aggregationMode,
      itemCount: Array.isArray(finalData) ? finalData.length : 'not array',
      sample: finalData?.[0]
    });

    return { data: finalData };
  }

  /**
   * Determine actual output data type for webSource nodes based on configuration
   */
  getDataType(nodeData: WorkflowNodeData): DataType {
    // Check monitoring config if present (this is where the actual config is stored)
    const monitoringConfig = nodeData.monitoring || nodeData;
    
    // Check if it's configured for table/structured output
    const selectors = monitoringConfig?.selectors || nodeData?.selectors;
    
    if (selectors?.some((s: any) => s.type === 'table' || s.type === 'repeating')) {
      console.log(`🔍 WebSource detected as STRUCTURED_DATA (table/repeating mode)`);
      return DataType.STRUCTURED_DATA;
    }
    
    // Check if it's configured and has multiple selectors (structured format)
    if (selectors?.length > 1) {
      console.log(`🔍 WebSource detected as STRUCTURED_DATA (multiple selectors)`);
      return DataType.STRUCTURED_DATA;
    }
    
    // Single text-based selectors output text data
    if (selectors?.length === 1) {
      const selector = selectors[0];
      if (selector.attribute === 'textContent' || selector.attribute === 'innerText') {
        console.log(`🔍 WebSource detected as TEXT_DATA (single text selector)`);
        return DataType.TEXT_DATA;
      }
    }
    
    // Check if configured
    if (selectors?.length > 0) {
      console.log(`🔍 WebSource detected as STRUCTURED_DATA (configured selectors)`);
      return DataType.STRUCTURED_DATA; // If configured, assume structured output
    }
    
    // Default to raw data if not specifically configured
    console.log(`🔍 WebSource detected as RAW_DATA (default/unconfigured)`);
    return DataType.RAW_DATA;
  }
}